src/server.rs

```diff
@@ -0,0 +1,409 @@
+#[no_mangle]
+pub extern "C" fn rustls_server_config_free(config: *const rustls_server_config) {
+    ffi_panic_boundary_unit! {
+        let config: &ServerConfig = try_ref_from_ptr!(config, &mut ServerConfig, ());
+        // To free the server_config, we reconstruct the Arc. It should have a refcount of 1,
+        // representing the C code's copy. When it drops, that refcount will go down to 0
+        // and the inner ServerConfig will be dropped.
+        let arc: Arc<ServerConfig> = unsafe { Arc::from_raw(config) };
+        let strong_count = Arc::strong_count(&arc);
+        if strong_count < 1 {
+            eprintln!(
+                "rustls_server_config_free: invariant failed: arc.strong_count was < 1: {}. \
+                You must not free the same server_config multiple times.",
+                strong_count
+            );
+        }
+    }
+}
+
+/// Create a new rustls::ServerSession, and return it in the output parameter `out`.
+/// If this returns an error code, the memory pointed to by `session_out` remains unchanged.
+/// If this returns a non-error, the memory pointed to by `session_out` is modified to point
+/// at a valid ServerSession. The caller now owns the ServerSession and must call
+/// `rustls_server_session_free` when done with it.
+#[no_mangle]
+pub extern "C" fn rustls_server_session_new(
+    config: *const rustls_server_config,
+    session_out: *mut *mut rustls_server_session,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let config: Arc<ServerConfig> = unsafe {
+            match (config as *const ServerConfig).as_ref() {
+                Some(c) => arc_with_incref_from_raw(c),
+                None => return NullParameter,
+            }
+        };
+        let server = ServerSession::new(&config);
+
+        // We've succeeded. Put the server on the heap, and transfer ownership
+        // to the caller. After this point, we must return CRUSTLS_OK so the
+        // caller knows it is responsible for this memory.
+        let b = Box::new(server);
+        unsafe {
+            *session_out = Box::into_raw(b) as *mut _;
+        }
+
+        return rustls_result::Ok;
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_server_session_wants_read(session: *const rustls_server_session) -> bool {
+    ffi_panic_boundary_bool! {
+        let session: &ServerSession = try_ref_from_ptr!(session, &ServerSession, false);
+        session.wants_read()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_server_session_wants_write(session: *const rustls_server_session) -> bool {
+    ffi_panic_boundary_bool! {
+        let session: &ServerSession = try_ref_from_ptr!(session, &ServerSession, false);
+        session.wants_write()
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rustls_server_session_is_handshaking(
+    session: *const rustls_server_session,
+) -> bool {
+    ffi_panic_boundary_bool! {
+        let session: &ServerSession = try_ref_from_ptr!(session, &ServerSession, false);
```

Comment:
nit/galaxy brain/definitely not worth trying: it feels like having to specify both `ffi_panic_boundary_bool!` and `false` is redundant, and also all these functions have a nearly identical preamble where you use `try_ref_from_ptr!` to get a `ServerSession`. I wonder if you could have a macro `server_ffi_panic_boundary_bool!` that you could use like so:

```
server_ffi_panic_boundary_bool! {
    session.is_handshaking()
}
```

I suspect I just have a bad case of macro madness, though.

Reply:
This does seem like macro madness! I'll pass for now but it's a cool idea. I bet it would work.